File: cogs/item_requests_scheduler.py
import time
import discord
import os
from discord.ext import commands, tasks
from config import (
    ITEM_REQUEST_SUMMARY_CHANNEL_ID
)

from utils.i18n import TEXT
import db

lang = "pt"
ITEM_IMAGES_DIR = "images/itens"


class ItemRequestsScheduler(commands.Cog):
    """
    Scheduler diário para cobrança automática de updates
    e controle de rank dos item requests.
    """

    def __init__(self, bot: commands.Bot):
        self.bot = bot
        self.daily_check.start()
        self.daily_loop.start()

    def cog_unload(self):
        self.daily_check.cancel()
        self.daily_loop.cancel()

    async def clear_summary_channel(self, channel: discord.TextChannel):
        await channel.purge(
            limit=100,
            check=lambda m: m.author == self.bot.user,
            bulk=True
        )

    async def send_item_summary(self):
        channel = self.bot.get_channel(ITEM_REQUEST_SUMMARY_CHANNEL_ID)
        if not channel:
            return

        await self.clear_summary_channel(channel)

        rows = db.get_daily_item_summary()
        if not rows:
            return

        embeds = {}
        files = {}

        for item_name, rank, player, remaining, thread_id in rows:
            if item_name not in embeds:
                embed = discord.Embed(
                    title=TEXT["rank_header"][lang].format(item=item_name),
                    color=discord.Color.blurple()
                )

                image_path = os.path.join(ITEM_IMAGES_DIR, f"{item_name}.png")

                if os.path.isfile(image_path):
                    file = discord.File(image_path, filename=f"{item_name}.png")
                    embed.set_thumbnail(url=f"attachment://{item_name}.png")
                    files[item_name] = file
                else:
                    files[item_name] = None

                embeds[item_name] = embed

            thread = self.bot.get_channel(thread_id)
            thread_link = thread.jump_url if thread else ""

            embeds[item_name].add_field(
                name=f"{rank}º - {player}",
                value=TEXT["rank_line"][lang].format(
                    remaining=remaining,
                    link=thread_link
                ),
                inline=False
            )

        for item_name, embed in embeds.items():
            file = files.get(item_name)
            if file:
                await channel.send(embed=embed, file=file)
            else:
                await channel.send(embed=embed)

    # ==========================================================
    # TASK DIÁRIA – RANK / WARN
    # ==========================================================

    @tasks.loop(minutes=60)
    async def daily_loop(self):
        await self.bot.wait_until_ready()
        await self.send_item_summary()

    @tasks.loop(hours=24)
    async def daily_check(self):
        await self.bot.wait_until_ready()

        now = time.time()
        requests = db.get_all_item_requests_for_check()

        for req in requests:
            (
                request_id,
                discord_id,
                player_name,
                item_name,
                rank_position,
                thread_id,
                thread_channel_id,
                last_update,
                warned_3d,
                warned_4d,
            ) = req

            if item_name in ("creature of gaiety", "elder dragon isteria", "carnival queen"):
                continue
            
            if rank_position != 1:
                db.fix_last_update(request_id, last_update)
                continue

            # ============================
            # 🔥 FAILSAFE TIMESTAMP
            # ============================

            # Se veio em ms
            if last_update > 10_000_000_000:
                last_update = int(last_update / 1000)
                db.fix_last_update(request_id, last_update)

            # Se veio do futuro
            if last_update > now:
                last_update = int(now)
                db.fix_last_update(request_id, last_update)

            days_idle = int((now - last_update) / 86400)

            thread = self.bot.get_channel(thread_id)
            if not thread:
                continue

            # =====================
            # 3 DIAS
            # =====================
            if days_idle >= 3 and not warned_3d:
                await thread.send(
                    TEXT["idle_3d"][lang].format(player=discord_id, item=item_name)
                )
                db.mark_request_warned(request_id, "warned_3d")
                continue

            # =====================
            # 4 DIAS
            # =====================
            if days_idle >= 4 and not warned_4d:
                await thread.send(
                    TEXT["idle_4d"][lang].format(player=discord_id, item=item_name)
                )
                db.mark_request_warned(request_id, "warned_4d")
                continue

            # =====================
            # 5 DIAS → RANK DOWN
            # =====================
            if days_idle >= 5:
                await thread.send(
                    TEXT["rank_down"][lang].format(player=discord_id, item=item_name)
                )
                db.drop_request_rank(request_id)

    # ==========================================================
    # START LOG
    # ==========================================================

    @daily_check.before_loop
    async def before_daily_check(self):
        await self.bot.wait_until_ready()
        print("[ItemRequestsScheduler] Scheduler diário iniciado.")


async def setup(bot: commands.Bot):
    await bot.add_cog(ItemRequestsScheduler(bot))

File: node/cogs/item_requests_scheduler.js
const fs = require('fs');
const path = require('path');
const { EmbedBuilder } = require('discord.js');
const { TEXT } = require('../utils/i18n');

const lang = 'pt';
const ITEM_IMAGES_DIR = path.join('images', 'itens');

async function clearSummaryChannel(channel, botUserId) {
  const messages = await channel.messages.fetch({ limit: 100 }).catch(() => null);
  if (!messages) {
    return;
  }

  const toDelete = messages.filter((message) => message.author.id === botUserId);
  if (toDelete.size > 0) {
    await channel.bulkDelete(toDelete, true).catch(() => null);
  }
}

async function sendItemSummary({ client, db, summaryChannelId }) {
  const channel = await client.channels.fetch(summaryChannelId).catch(() => null);
  if (!channel) {
    return;
  }

  await clearSummaryChannel(channel, client.user.id);

  const rows = await db.getDailyItemSummary();
  if (!rows || rows.length === 0) {
    return;
  }

  const embeds = new Map();
  const files = new Map();

  for (const row of rows) {
    if (!embeds.has(row.item_name)) {
      const embed = new EmbedBuilder()
        .setTitle(TEXT.rank_header[lang].replace('{item}', row.item_name))
        .setColor(0x5865f2);

      const imagePath = path.join(ITEM_IMAGES_DIR, `${row.item_name}.png`);
      if (fs.existsSync(imagePath)) {
        embed.setThumbnail(`attachment://${row.item_name}.png`);
        files.set(row.item_name, imagePath);
      }

      embeds.set(row.item_name, embed);
    }

    const thread = await client.channels.fetch(row.thread_id).catch(() => null);
    const link = thread?.url || '';

    embeds.get(row.item_name).addFields({
      name: `${row.rank_position}º - ${row.player_name}`,
      value: TEXT.rank_line[lang]
        .replace('{remaining}', row.remaining_quantity)
        .replace('{link}', link),
      inline: false,
    });
  }

  for (const [itemName, embed] of embeds.entries()) {
    const filePath = files.get(itemName);
    if (filePath) {
      await channel.send({
        embeds: [embed],
        files: [{ attachment: filePath, name: `${itemName}.png` }],
      });
    } else {
      await channel.send({ embeds: [embed] });
    }
  }
}

function startItemRequestScheduler({ client, db, summaryChannelId }) {
  const hourlyInterval = setInterval(async () => {
    await sendItemSummary({ client, db, summaryChannelId });
  }, 60 * 60 * 1000);

  const dailyInterval = setInterval(async () => {
    const now = Math.floor(Date.now() / 1000);
    const requests = await db.getAllItemRequestsForCheck();

    for (const req of requests) {
      if (req.item_name === 'creature of gaiety') {
        continue;
      }
      if (req.item_name === 'elder dragon isteria') {
        continue;
      }

      const daysIdle = Math.floor((now - req.last_update) / 86400);
      const thread = await client.channels.fetch(req.thread_id).catch(() => null);
      if (!thread) {
        continue;
      }

      if (daysIdle >= 3 && !req.warned_3d) {
        await thread.send(
          TEXT.idle_3d[lang]
            .replace('{player}', req.discord_id)
            .replace('{item}', req.item_name),
        );
        await db.markRequestWarned(req.id, 'warned_3d');
        continue;
      }

      if (daysIdle >= 4 && !req.warned_4d) {
        await thread.send(
          TEXT.idle_4d[lang]
            .replace('{player}', req.discord_id)
            .replace('{item}', req.item_name),
        );
        await db.markRequestWarned(req.id, 'warned_4d');
        continue;
      }

      if (daysIdle >= 5) {
        await thread.send(
          TEXT.rank_down[lang]
            .replace('{player}', req.discord_id)
            .replace('{item}', req.item_name),
        );
        await db.dropRequestRank(req.id);
      }
    }
  }, 24 * 60 * 60 * 1000);

  return () => {
    clearInterval(hourlyInterval);
    clearInterval(dailyInterval);
  };
}

module.exports = {
  sendItemSummary,
  startItemRequestScheduler,
};
